### PR TITLE
Implement rev query parameter for HEAD, GET, PUT and DELETE requests.

### PR DESCRIPTION
**Pull request by jeqq**

### CODE REVIEW

**Comment by dickolsson** [comment]
@jeqq Looks good!

Let's also do the same for GET and HEAD requests and add tests for that, e.g. 404 should be returned if the revision doesn't exist.

**Comment by dickolsson** [comment]
And for `DELETE` if the revision exists but is _not_ the _latest_ revision, we should return a `409 Conflict` HTTP status (`throw new \Symfony\Component\HttpKernel\Exception\ConflictHttpException`). The reason for this is that one should not be able to delete old revisions.

See http://docs.couchdb.org/en/latest/api/document/common.html#delete--db-docid

**Comment by dickolsson** [comment]
And while at it, let's do the same for the `PUT` method. We need to return `409 Conflict` if the `rev` parameter is not the latest revision. The reason for this is that it should not be possible to update old revisions.s

File: src/Plugin/rest/resource/DocResource.php [hunk 1]
```diff
@@ -132,12 +138,20 @@ public function delete($workspace, $entity) {
     if (!$entity->access('delete')) {
       throw new AccessDeniedHttpException();
     }
+
+    $last_entity_revision = entity_load($entity->getEntityType()->id(), $entity->id());
```
Comment (dickolsson) [inline src/Plugin/rest/resource/DocResource.php]: Instead of using `entity_load()` it's better if we add the `rev` key to `\Drupal\multiversion\Entity\UuidIndex::buildValue`. Because then we can do this instead, which is MUCH faster:

```
$record = \Drupal::service('entity.uuid_index')->get($entity->uuid());
$last_rev = $record['rev'];
```

So please update `\Drupal\multiversion\Entity\UuidIndex::buildValue` and add `rev` (from `$entity->_revs_info->rev`) and `revision_id` (from `$entity->getRevisionId()`).

Then we also need to make sure that the UUID index is being kept up to date **all the time** (insert, update and delete), because right now it's only update on `hook_entity_insert` and `hook_entity_delete` in `multiversion.module`.

Comment (dickolsson) [inline src/Plugin/rest/resource/DocResource.php]: While we're at it, please add `local_seq` to `UuidIndex` as well.

And add `rev` key to `RevisionIndex` as well. Then both indexes contain the same information which good for consistency.

File: src/Tests/DocResourceTest.php [hunk 2]
```diff
@@ -123,6 +127,22 @@ public function testPut() {
...
+      $entity = entity_create($entity_type);
+      $entity->save();
+      $first_rev = $entity->_revs_info->rev;
+      $entity->name = $this->randomMachineName();
+      $entity->save();
+      $second_rev = $entity->_revs_info->rev;
+      $serialized = $serializer->serialize($entity, $this->defaultFormat);
+
+      $this->httpRequest("$db/" . $entity->uuid(), 'PUT', $serialized, NULL, array('if-match' => $first_rev));
+      $this->assertResponse('409', 'HTTP response code is correct.');
```
Comment (dickolsson) [inline src/Tests/DocResourceTest.php:171]: Really nice! Please also add a test that the same `409` result is also given when using `?rev=` query parameter instead.

File: src/Tests/DocResourceTest.php [hunk 3]
```diff
@@ -123,6 +127,22 @@ public function testPut() {
...
+
+      $response = $this->httpRequest("$db/" . $entity->uuid(), 'PUT', $serialized, NULL, array('if-match' => $second_rev));
+      $this->assertResponse('201', 'HTTP response code is correct.');
```
Comment (dickolsson) [inline src/Tests/DocResourceTest.php:174]: Nice, this is awesome!

File: src/Tests/DocResourceTest.php [hunk 4]
```diff
@@ -148,7 +168,25 @@ public function testDelete() {
...
+
+      $entity = entity_create($entity_type);
+      $entity->save();
+      $first_rev = $entity->_revs_info->rev;
+      $entity->name = $this->randomMachineName();
+      $entity->save();
+      $second_rev = $entity->_revs_info->rev;
+
+      $this->httpRequest("$db/" . $entity->uuid(), 'DELETE', array('rev' => $first_rev));
+      $this->assertResponse('409', 'HTTP response code is correct.');
```
Comment (dickolsson) [inline src/Tests/DocResourceTest.php:221]: Please also add a test that does the same with the `if-match` header. It should have the same `409` result.

File: src/ParamConverter/DocIdConverter.php [hunk 5]
```diff
@@ -51,7 +51,9 @@ public function convert($uuid, $definition, $name, array $defaults) {
     // Fetch parameters.
     $open_revs_query = trim($request->query->get('open_revs'), '[]');
     if (!$rev_query = $request->query->get('rev')) {
-      $rev_query = $request->headers->get('if-none-match');
+      if (!$rev_query = $request->headers->get('if-none-match')) {
```
Comment (dickolsson) [inline src/ParamConverter/DocIdConverter.php:54]: Please add a test that does `GET` and `HEAD` with `if-none-match` header to ensure the right revision is returned.

File: src/Tests/DocResourceTest.php (hunk 2 above)
Comment (jeqq) [inline src/Tests/DocResourceTest.php:171]: @dickolsson For GET, HEAD and DELETE methods we don't have CURLOPT_POSTFIELDS option and the query parameters are added in the $body method parameter (and then is used in CURLOPT_URL option), but for the PUT method we have the CURLOPT_POSTFIELDS and also we need to add the rev query parameter. My propose is to add a new method parameter:

```
httpRequest($url, $method, $body = NULL, $mime_type = NULL, $headers = NULL, $parameters = NULL)
```

In this case, in all tests the httpRequest() should be modified. What do you think?

Comment (dickolsson) [inline src/Tests/DocResourceTest.php:171]: I think that sounds good. I would just rename `$parameters` to `$query`, to make it more clear.
